### PR TITLE
Add unused diagnostic subtype

### LINKED ISSUE #15710: Highlight unused symbols
I'd like to request an improvement to Typescript support: highlighting of unused symbols.

Refactoring and moving things around between files often leads to a lot of manual analysis, for example to find imports that are no longer in use.

This doesn't need to be a big, complex feature - I don't expect this to trawl through the entire codebase and figure out what is or isn't in use.

I'd be happy with just file-local highlighting - e.g. gray out variable and constant declarations if they aren't exported or used in any expression or statement anywhere in the file.

### PR DESCRIPTION
**Pull request by mjbvz**
Fixes #15710

Adds a new `DiagnosticTag` class that provide additional information about a diagnostic. Introduce `DiagnosticTag.Unnecessary` to mark when a diagnostic is for unused / unnecessary code

The design comes from Rosyln's diagnostic object and allows us to modify how a diagnostic is rendered without changing its serverity.

Hooks up JS and TS to use this new tag. This is controlled by the `javascript.showUnused.enabled` setting which is enabled by default

## Alternate design considered

- Introduce a new diagnostic severity for unused.
    
    However, using this approach, if a user sets `noUnusedLocals` in their `tsconfig.json`, the resulting diagnostic could only show the squiggly OR be grayed out. Using `customTags` allows us to support both graying out and showing the squiggly

- Custom JS/TS implementation using decorators

    Not themable. We want a standard experience across languages.

### CODE REVIEW

**Comment by mjbvz** [comment]
Other tags the Roslyn supports: http://source.roslyn.io/#Microsoft.CodeAnalysis/Diagnostic/WellKnownDiagnosticTags.cs,03ef56ba68ffca36

File: src/vs/vscode.d.ts [hunk 1]
```diff
@@ -3844,6 +3844,16 @@ declare module 'vscode' {
...
 	}
 
+	/**
+	 * Additional metadata about the type of diagnostic.
+	 */
+	export enum DiagnosticTag {
+		/**
+		 * Unused or unnecessary code.
+		 */
+		Unnecessary = 'unnecessary'
```
Comment (jrieken) [inline src/vs/vscode.d.ts]: https://github.com/Microsoft/vscode/wiki/Extension-API-guidelines#enums

**Review by jrieken — CHANGES_REQUESTED** [review]

**Comment by jrieken** [comment]
What other tags do we envision? Any of those used by Roslyn seem very specific: http://source.roslyn.io/#Microsoft.CodeAnalysis/Diagnostic/WellKnownDiagnosticTags.cs,03ef56ba68ffca36,references, in fact I only understand half of them... Is this maybe a dupe of https://github.com/Microsoft/vscode/issues/24682 just with some predefined tags/defaults that we know how to render? Will there be a future with custom rendering of tags?

**Comment by jrieken** [comment]
Further questions to seed some disucssion

* Why do we like TSLint using green squiggles for unused things but don't like that for TypeScript?
* Do we want `DiagnosticSeverity.Hidden` to help code actions but do the rendering via some (tbd) semantic highlights API?

**Comment by mjbvz** [comment]
Merging with proposed api using diagnostic tags. We can revisit this to see if we want try implementing this with a semantic highlighting api next iteration

**Pull request merged**